daemon/emer-aggregate-tally.c

```diff
@@ -201,37 +252,133 @@ emer_aggregate_tally_constructed (GObject *object)
+                                "    aggregate_key,\n"
+                                "    payload\n"
+                                ")",
+                                NULL, NULL, NULL)) ||
+          !CHECK (sqlite3_exec (self->db, "PRAGMA user_version = 1", NULL, NULL, NULL)))
+        return FALSE;
+      /* fallthrough: */
+    case 1:
+      break;
+    case -1:
+      return FALSE;
+    default:
+      g_set_error (error, EMER_SQLITE_ERROR, SQLITE_ERROR,
+                   "Unexpected user_version %" G_GINT64_FORMAT,
+                   (gint64) user_version);
+      return FALSE;
+    }
+
+  return TRUE;
 }
 
 static void
-emer_aggregate_tally_finalize (GObject *object)
+emer_aggregate_tally_delete_db (EmerAggregateTally *self,
+                                const char         *db_path)
 {
-  EmerAggregateTally *self = (EmerAggregateTally *)object;
+  /*
+   * When in active use, the state of a WAL mode database is described by
+   * three separate files:
+   *
+   * 1. The main database file with an arbitrary name "X".
+   * 2. The write-ahead log file, usually named "X-wal".
+   * 3. The wal-index file, usually named "X-shm".
+   *
+   *  -- https://sqlite.org/walformat.html
+   */
+  const gchar *suffixes[] = {
+    "",
+    "-shm",
+    "-wal"
+  };
+
+  for (size_t i = 0; i < G_N_ELEMENTS (suffixes); i++)
+    {
+      g_autofree gchar *path = g_strconcat (db_path, suffixes[i], NULL);
+
+      if (0 != g_unlink (path) && errno != ENOENT)
+        g_warning ("Failed to unlink %s: %s", path, g_strerror (errno));
+    }
+}
+
+static void
+emer_aggregate_tally_constructed (GObject *object)
+{
+  EmerAggregateTally *self = EMER_AGGREGATE_TALLY (object);
+  g_autofree gchar *old_path = NULL;
+  g_autofree gchar *path = NULL;
+  g_autoptr(GError) error = NULL;
+
+  G_OBJECT_CLASS (emer_aggregate_tally_parent_class)->constructed (object);
+
+  g_assert (self->persistent_cache_directory != NULL);
+  ensure_folder_exists (self, self->persistent_cache_directory, NULL);
+
+ /* No daemon version that used this path made it into a production release,
+  * or even beta. So rather than migrating the data for the schema & filename
+  * change, just delete it (and ignore errors).
+  */
+  old_path = g_build_filename (self->persistent_cache_directory,
+                               "aggregate-events.db",
+                               NULL);
+  emer_aggregate_tally_delete_db (self, old_path);
 
-  if (self->db)
+  path = g_build_filename (self->persistent_cache_directory,
+                           "metrics.db",
+                           NULL);
+  if (!emer_aggregate_tally_init_db (self, path, &error))
     {
-      CHECK (sqlite3_close (self->db));
-      self->db = NULL;
+      g_warning ("Failed to initialize %s: %s; trying to delete and recreate it",
+                 path, error->message);
+      g_clear_error (&error);
+
+      g_clear_pointer (&self->db, close_db);
+      emer_aggregate_tally_delete_db (self, path);
```

Comment:
I realised while reworking this that it is necessary to delete 3 files, not 1. Happily I had already written that code to delete the old named DB.